src/io/simulator/simulator_handle.hpp

```diff
@@ -0,0 +1,453 @@
+      if (dop.deadline < now) {
+        std::cout << "timing out request" << std::endl;
+        DeadlineAndOpaquePromise dop = std::move(promises_.at(promise_key));
+        promises_.erase(promise_key);
+
+        stats_.timed_out_requests++;
+
+        dop.promise.TimeOut();
+      }
+    }
+  }
+
+  bool MaybeTickSimulator() {
+    std::unique_lock<std::mutex> lock(mu_);
+
+    size_t blocked_servers = blocked_on_receive_;
+
+    for (auto &[promise_key, opaque_promise] : promises_) {
+      if (opaque_promise.promise.IsAwaited()) {
+        if (server_addresses_.contains(promise_key.requester_address)) {
+          blocked_servers++;
+        }
+      }
+    }
+
+    if (blocked_servers < server_addresses_.size()) {
+      // we only need to advance the simulator when all
+      // servers have reached a quiescent state, blocked
+      // on their own futures or receive methods.
+      return false;
+    }
+
+    stats_.simulator_ticks++;
+
+    cv_.notify_all();
+
+    TimeoutPromisesPastDeadline();
+
+    if (in_flight_.empty()) {
+      // return early here because there are no messages to schedule
+
+      // We tick the clock forward when all servers are blocked but
+      // there are no in-flight messages to schedule delivery of.
+      std::poisson_distribution<> time_distrib(50);
+      uint64_t clock_advance = time_distrib(rng_);
+      cluster_wide_time_microseconds_ += clock_advance;
+
+      MG_ASSERT(cluster_wide_time_microseconds_ < config_.abort_time,
+                "Cluster has executed beyond its configured abort_time, and something may be failing to make progress "
+                "in an expected amount of time.");
+      return true;
+    }
+
+    if (config_.scramble_messages) {
+      // scramble messages
+      std::uniform_int_distribution<size_t> swap_distrib(0, in_flight_.size() - 1);
+      size_t swap_index = swap_distrib(rng_);
+      std::swap(in_flight_[swap_index], in_flight_.back());
+    }
+
+    auto [to_address, opaque_message] = std::move(in_flight_.back());
+    in_flight_.pop_back();
+
+    std::uniform_int_distribution<int> drop_distrib(0, 99);
+    int drop_threshold = drop_distrib(rng_);
+    bool should_drop = drop_threshold < config_.drop_percent;
+
+    if (should_drop) {
+      stats_.dropped_messages++;
+    }
+
+    PromiseKey promise_key{.requester_address = to_address,
+                           .request_id = opaque_message.request_id,
+                           .replier_address = opaque_message.from_address};
+
+    if (promises_.contains(promise_key)) {
+      // complete waiting promise if it's there
+      DeadlineAndOpaquePromise dop = std::move(promises_.at(promise_key));
+      promises_.erase(promise_key);
+
+      bool normal_timeout = config_.perform_timeouts && (dop.deadline < cluster_wide_time_microseconds_);
+
+      if (should_drop || normal_timeout) {
+        stats_.timed_out_requests++;
+        dop.promise.TimeOut();
```

Comment:
So this means if a message is dropped, that can cause a promise to timeout before the timeout of the request, right? I find this tricky, because in real world dropping would result in some network error like "disconnected" or "connection is broken" or something like that.

Reply:
What is your opinion here?

Reply:
I think the more chaos introduced in terms of timing, the better. Any code that depends on a certain duration of time for correctness has a bug.  To this end, the simulator will eventually move clocks backwards as well, getting buggy math related to timeout handling to jump out. This is "real" because NTP can do the same thing if the delta is above a few minutes. 

But I think in terms of realism, there should only be one place that times out requests, and that should be in the TimeoutPromisesPastDeadline method. I'll make this specific case a no-op and just keep the associated promise in the promises_ map until timeouts get dealt with in the single path.